### DISCUSSION TITLE
How should I trigger force rebuild of module using webpack plugin api?

### DISCUSSION THREAD

**Discussion by artem1458** [webpack/webpack#13528] [General]
Hi!
I'm writing plugin for webpack, and I want to trigger rebuild of FileB when FileA was changed, FIleA and FileB are not linked at the dependency level
I'm trying to call compilation.rebuildModule with module FileB found in compilation.modules, and it works. After ~1min of not changing any files, and after trying to change FileA (to trigger compilation.rebuildModule), build fails with this stacktrace
Maybe I need to call rebuild on another stage (not on buildModule hook)
Can someone please help me?

Here is simplified code of plugin
```ts
compiler.hooks.compilation.tap('Compiler Compilation hook', (compilation) => {
            compilation.hooks.buildModule.tap('Rebuild ModuleB when ModuleA changed', (module) => {
                const currentBuiltModule = module as NormalModule;

                if (!currentBuiltModule.resource || /FileA\.ts$/.test(currentBuiltModule.resource)) {
                    return;
                }

                const webpackModule = Array.from(compilation.modules).find(it =>
                    it.resource && /FileB\.ts$/.test(it.resource)
                );

                if (webpackModule) {
                    compilation.rebuildModule(webpackModule, (error, result) => {
                        error && compilation.errors.push(error);
                    });
                }
            });
        });

```

Stacktrace with error
```
TypeError [ERR_INVALID_ARG_TYPE]: The "data" argument must be of type string or an instance of Buffer, TypedArray, or DataView. Received undefined
    at Hash.update (internal/crypto/hash.js:84:11)
    at BulkUpdateDecorator.update (node_modules/webpack/lib/util/createHash.js:52:15)
    at NormalModule.updateHash (node_modules/webpack/lib/NormalModule.js:1272:8)
    at Compilation._createModuleHash (node_modules/webpack/lib/Compilation.js:3213:10)
    at Compilation.createModuleHashes (node_modules/webpack/lib/Compilation.js:3185:10)
    at node_modules/webpack/lib/Compilation.js:2447:11
    at Hook.eval [as callAsync] (eval at create (node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (node_modules/tapable/lib/Hook.js:18:14)
    at node_modules/webpack/lib/Compilation.js:2407:36
    at eval (eval at create (node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:11:1)
```

**Comment by alexander-akait** [comment]
For questions, please use `Discussions`. What do you try achieve? Why do you need rebuild module?

**Reply by artem1458** [reply]
Hi @alexander-akait. Thanks for the reply :slightly_smiling_face: 
I'm developing DI-Container that automatically resolves the dependencies of classes by their type (DI for typescript). So, when the file with class is changing, I need to call rebuild of file in which described DI config, to re-resolve dependencies

I've created reproducible repo https://github.com/artem1458/webpack-bug-reproduce

**Reply by alexander-akait** [reply]
It will be very bad for performance. Each rebuild increases build time very very very well.

**Reply by artem1458** [reply]
Yeah, but shouldn't it be the same as an editing file? :slightly_smiling_face:

**Reply by artem1458** [reply]
Because this thing with rebuilding is only for hmr

**Reply by alexander-akait** [reply]
Maybe related https://github.com/webpack/webpack/issues/10392 and here example how you can use it https://github.com/webpack/webpack/pull/11433/files

**Reply by artem1458** [reply]
@alexander-akait Thanks for targeting :slightly_smiling_face: 
I've solved this issue with the example you shared.
Thnx a lot!
Later I will share code with example for multiple files here.

**Comment by baiwusanyu-c** [comment]
Hello, you finally solved, can you provide me with a case code? I have similar scenes with you. I want to re-compile the Vue single file component after changing the CSS file, because I will modify the `vue-loader` code after the conversion phase, which will eventually affect the output of the page, and the content of these modified content is based on CSS file content. I don't know what to do now
@artem1458

**Reply by artem1458** [reply]
Hey, @baiwusanyu-c here is plugin that is rebuild some dependent files even if code in files is not changed 
Write to me if you need any additional info or help 🙂 
[https://github.com/artem1458/dependency-injection-cat/blob/main/src/plugins/webpack/index.ts](https://github.com/artem1458/dependency-injection-cat/blob/main/src/plugins/webpack/index.ts)

**Reply by baiwusanyu-c** [reply]
Thank you so much! hahahahaha! I solved it, this has been bugging me for days, thanks for your help! !
